core/src/Kokkos_Get.hpp

```diff
@@ -0,0 +1,147 @@
+/*
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 2.0
+//              Copyright (2019) Sandia Corporation
+//
+// Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_CORE_GET_HPP
+#define KOKKOS_CORE_GET_HPP
+
+#include <Properties/Kokkos_Detection.hpp>
+
+#include <utility>  // std::get
+
+namespace Kokkos {
+
+//==============================================================================
+// <editor-fold desc="get Niebloid implementation details"> {{{1
+
+namespace Impl {
+namespace _get_impl_ignore_adl {
+
+// Poison-pill overload
+template <class T, size_t I>
+void get(T&&) = delete;
+
+KOKKOS_DECLARE_DETECTION_ARCHETYPE(_has_adl_free_function_get_archetype,
+                                   (class T, class IType), (T, IType),
+                                   decltype(get<IType::value>(declval<T>())));
+template <class T, size_t I>
+using has_adl_free_function_get =
+    Kokkos::Impl::is_detected<_has_adl_free_function_get_archetype, T,
+                              std::integral_constant<size_t, I>>;
+
+KOKKOS_DECLARE_DETECTION_ARCHETYPE(
+    _has_std_get_archetype, (class T, class IType), (T, IType),
```

Comment:
where is this "_" naming scheme coming from. I don't think anything else in Kokkos does that?

Reply:
I use the underscore prefix for things that are implementation details of something already in namespace `Impl`. I find the lack of any convention for this practice makes Kokkos code hard to read at times, so I've tried to start one.  If you have other ideas for such a convention, I'm happy to use that instead.

Reply:
I believe we resolved this at the meeting, I think this pattern is fine to _use_, we still need to decide whether to enforce it